cli/src/package.rs

```diff
@@ -0,0 +1,58 @@
+extern crate tar;
+extern crate ignore;
+use std::path::Path;
+use ignore::Walk;
+use std::fs::File;
+use tar::Builder;
+use tempfile::tempfile;
+use flate2::Compression;
+use std::io::BufReader;
+use flate2::bufread::GzEncoder;
+use std::io::{Write, Read, SeekFrom, Seek};
+
+pub fn create_package(path: String, output_path: String) {
+  // note(itay): The ignore crate uses the .gitignore file and also a .ignore file
+  // if specified by default, so basically like walkdir but already baked with the
+  // logic we had in mind.
+  let files_to_package = ignore::Walk::new(path);
+  let tar = create_tar(files_to_package);
+  let compressed_tar = get_compressed_tar(tar);
+  save_buffer_to_file(compressed_tar, output_path.to_string());
+} 
+
+fn create_tar(files_to_tar: Walk) -> File {
+  let mut tar_file = tempfile().unwrap();
+  {
+    let mut tar_builder = Builder::new(&tar_file);
+    for file in files_to_tar {
+      let entry = file.unwrap();
+      // println!("{}", entry.path().display());
+      tar_builder.append_path(entry.path()).unwrap();
+    }
+    tar_builder.finish().unwrap();
+  }
+  &mut tar_file.seek(SeekFrom::Start(0)).unwrap();
+  return tar_file;
+}
+
+fn get_compressed_tar(tar: File) -> Vec<u8> {
+  let buf_reader = BufReader::new(tar);
+  let mut encoder = GzEncoder::new(buf_reader, Compression::default());
+  let mut buffer = Vec::new();
+  encoder.read_to_end(&mut buffer).unwrap();
+  return buffer;
+}
+
+fn save_buffer_to_file(buffer: Vec<u8>, path: String) {
+  let mut output_file = File::create(path).unwrap();
+  output_file.write_all(&buffer).unwrap();
+}
+
+#[test]
+fn should_create_package() {
+  let output_path = "/tmp/test_package.tar.gz";
+  create_package("./tests/fixtures/test_package".to_string(), output_path.to_string());
```

Comment:
I do to_string() everywhere but I need to read about strings to see what the fuck is going on with it

Reply:
yes, please do and tell me..

Reply:
changed to &str
https://hermanradtke.com/2015/05/03/string-vs-str-in-rust-functions.html